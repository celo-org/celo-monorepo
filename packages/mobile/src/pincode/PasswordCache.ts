const CACHE_TIMEOUT = 300000 // 5 minutes

interface SecretCache {
  [account: string]: {
    timestamp: number | null
    secret: string | null
  }
}
let pinCache: SecretCache = {}
let pepperCache: SecretCache = {}
let passwordHashCache: SecretCache = {}
let passwordCache: SecretCache = {}

function getCachedValue(cache: SecretCache, account: string) {
  // TODO(Rossy) use a monotonic clock here?
  const value = cache[account]
  if (value && value.secret && value.timestamp && Date.now() - value.timestamp < CACHE_TIMEOUT) {
    return value.secret
  } else {
    // Clear values in cache when they're expired
    cache[account] = { timestamp: null, secret: null }
    return null
  }
}

function setCachedValue(cache: SecretCache, account: string, secret: string | null) {
  if (!cache[account]) {
    cache[account] = { timestamp: null, secret: null }
  }
  cache[account].timestamp = Date.now()
  cache[account].secret = secret
}

export function getCachedPin(account: string) {
  return getCachedValue(pinCache, account)
}

export function setCachedPin(account: string, pin: string | null) {
  setCachedValue(pinCache, account, pin)
}

export function getCachedPepper(account: string) {
  return getCachedValue(pepperCache, account)
}

export function setCachedPepper(account: string, pepper: string | null) {
  setCachedValue(pepperCache, account, pepper)
}

export function getCachedPasswordHash(account: string) {
  return getCachedValue(passwordHashCache, account)
}

export function setCachedPasswordHash(account: string, passwordHash: string) {
  setCachedValue(passwordHashCache, account, passwordHash)
}

export function getCachedPassword(account: string) {
  return getCachedValue(passwordCache, account)
}

export function setCachedPassword(account: string, password: string | null) {
  setCachedValue(passwordCache, account, password)
}

export function clearPasswordCaches() {
  pinCache = {}
  pepperCache = {}
  passwordHashCache = {}
  passwordCache = {}
}
