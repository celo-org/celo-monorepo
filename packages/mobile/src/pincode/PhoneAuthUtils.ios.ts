export function isPhoneAuthSupported() {
  return false
}

export async function setPinInKeystore() {
  throw new Error('Not supported')
}

export async function getPinFromKeystore() {
  throw new Error('Not supported')
}
