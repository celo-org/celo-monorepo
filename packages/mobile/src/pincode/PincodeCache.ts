const PIN_TIMEOUT = 300000 // 5 minutes

interface PincodeCache {
  timestamp: number | null
  pincode: string | null
}

const pincodeCache: PincodeCache = {
  timestamp: null,
  pincode: null,
}

export function getCachedPincode() {
  // TODO(Rossy) use a  monotonic clock here
  if (
    pincodeCache.pincode &&
    pincodeCache.timestamp &&
    Date.now() - pincodeCache.timestamp < PIN_TIMEOUT
  ) {
    return pincodeCache.pincode
  }
  return null
}

export function setCachedPincode(pincode: string) {
  pincodeCache.timestamp = Date.now()
  pincodeCache.pincode = pincode
}
