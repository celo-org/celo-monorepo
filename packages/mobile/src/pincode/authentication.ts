/**
 * Logic and utilities for managing account secrets
 * The pincode is a short numeric string the user is required to enter
 * The pepper is a generated once per account and stored in the keychain/keystore
 * The password is a combination of the two. It is used for unlocking the account in geth
 */

import { isValidAddress, normalizeAddress } from '@celo/utils/src/address'
import { sha256 } from 'ethereumjs-util'
import { asyncRandomBytes } from 'react-native-secure-randombytes'
import { call, select } from 'redux-saga/effects'
import { PincodeType } from 'src/account/reducer'
import { pincodeTypeSelector } from 'src/account/selectors'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { UNLOCK_DURATION } from 'src/geth/consts'
import i18n from 'src/i18n'
import { navigate, navigateBack } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import {
  clearPasswordCaches,
  getCachedPassword,
  getCachedPasswordHash,
  getCachedPepper,
  getCachedPin,
  setCachedPassword,
  setCachedPasswordHash,
  setCachedPepper,
  setCachedPin,
} from 'src/pincode/PasswordCache'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { getWalletAsync } from 'src/web3/contracts'

const TAG = 'pincode/authentication'

enum STORAGE_KEYS {
  PEPPER = 'PEPPER',
  PASSWORD_HASH = 'PASSWORD_HASH',
}

const PEPPER_LENGTH = 64
export const PIN_LENGTH = 6
// Pepper and pin not currently generalized to be per account
// Using this value in the caches
export const DEFAULT_CACHE_ACCOUNT = 'default'

const PIN_BLACKLIST = [
  '000000',
  '111111',
  '222222',
  '333333',
  '444444',
  '555555',
  '666666',
  '777777',
  '888888',
  '999999',
  '123456',
  '654321',
]

export function isPinValid(pin: string) {
  return pin.length === PIN_LENGTH && !PIN_BLACKLIST.includes(pin)
}

export async function retrieveOrGeneratePepper() {
  if (!getCachedPepper(DEFAULT_CACHE_ACCOUNT)) {
    let storedPepper = await retrieveStoredItem(STORAGE_KEYS.PEPPER)
    if (!storedPepper) {
      const randomBytes = await asyncRandomBytes(PEPPER_LENGTH)
      const pepper = randomBytes.toString('hex')
      await storeItem({ key: STORAGE_KEYS.PEPPER, value: pepper })
      storedPepper = pepper
    }
    setCachedPepper(DEFAULT_CACHE_ACCOUNT, storedPepper)
  }
  return getCachedPepper(DEFAULT_CACHE_ACCOUNT)!
}

async function getPasswordForPin(pin: string) {
  const pepper = await retrieveOrGeneratePepper()
  const password = `${pepper}${pin}`
  return password
}

async function getPasswordHashForPin(pin: string) {
  const password = await getPasswordForPin(pin)
  return getPasswordHash(password)
}

function getPasswordHash(password: string) {
  return sha256(new Buffer(password, 'hex')).toString('hex')
}

function passwordHashStorageKey(account: string) {
  if (!isValidAddress(account)) {
    throw new Error('Expecting valid address for computing storage key')
  }
  return `${STORAGE_KEYS.PASSWORD_HASH}-${normalizeAddress(account)}`
}

function storePasswordHash(hash: string, account: string) {
  setCachedPasswordHash(account, hash)
  return storeItem({ key: passwordHashStorageKey(account), value: hash })
}

async function retrievePasswordHash(account: string) {
  if (!getCachedPasswordHash(account)) {
    let hash: string | null = null
    try {
      hash = await retrieveStoredItem(passwordHashStorageKey(account))
    } catch (err) {
      Logger.error(`${TAG}@retrievePasswordHash`, 'Error retrieving hash', err, true)
      return null
    }
    if (!hash) {
      Logger.warn(`${TAG}@retrievePasswordHash`, 'No password hash found in store')
      return null
    }
    setCachedPasswordHash(account, hash)
  }
  return getCachedPasswordHash(account)
}

export async function getPassword(
  account: string,
  withVerification: boolean = true,
  storeHash: boolean = false
) {
  let password = getCachedPassword(account)
  if (password) {
    return password
  }

  const pin = await getPincode(withVerification)
  password = await getPasswordForPin(pin)

  if (storeHash) {
    const hash = getPasswordHash(password)
    await storePasswordHash(hash, account)
  }

  setCachedPassword(account, password)
  return password
}

export function* getPasswordSaga(account: string, withVerification?: boolean, storeHash?: boolean) {
  const pincodeType = yield select(pincodeTypeSelector)

  if (pincodeType === PincodeType.Unset) {
    Logger.error(TAG + '@getPincode', 'Pin has never been set')
    ValoraAnalytics.track(OnboardingEvents.pin_never_set)
    throw Error('Pin has never been set')
  }

  if (pincodeType !== PincodeType.CustomPin) {
    throw new Error(`Unsupported Pincode Type ${pincodeType}`)
  }

  return yield call(getPassword, account, withVerification, storeHash)
}

type PinCallback = (pin: string) => void

// Retrieve the pincode value
// May trigger the pincode enter screen
export async function getPincode(withVerification = true) {
  const cachedPin = getCachedPin(DEFAULT_CACHE_ACCOUNT)
  if (cachedPin) {
    return cachedPin
  }

  const pin = await requestPincodeInput(withVerification, true)
  return pin
}

// Navigate to the pincode enter screen and check pin
export async function requestPincodeInput(withVerification = true, shouldNavigateBack = true) {
  const pin = await new Promise((resolve: PinCallback) => {
    navigate(Screens.PincodeEnter, {
      onSuccess: resolve,
      withVerification,
    })
  })

  if (shouldNavigateBack) {
    navigateBack()
  }

  if (!pin) {
    throw new Error('Pincode confirmation returned empty pin')
  }

  setCachedPin(DEFAULT_CACHE_ACCOUNT, pin)
  return pin
}

// Confirm pin is correct by checking it against the stored password hash
export async function checkPin(pin: string, account: string) {
  const hashForPin = await getPasswordHashForPin(pin)
  const correctHash = await retrievePasswordHash(account)

  if (!correctHash) {
    Logger.warn(`${TAG}@checkPin`, 'No password hash stored. Checking with rpcWallet instead.')
    const password = await getPasswordForPin(pin)
    const unlocked = await ensureCorrectPassword(password, account)
    if (unlocked) {
      await storePasswordHash(hashForPin, account)
      return true
    }
    return false
  }

  return hashForPin === correctHash
}

// Confirm password by actually attempting to unlock the account
export async function ensureCorrectPassword(
  password: string,
  currentAccount: string
): Promise<boolean> {
  try {
    const wallet = await getWalletAsync()
    const result = await wallet.unlockAccount(currentAccount, password, UNLOCK_DURATION)
    return result
  } catch (error) {
    Logger.error(TAG, 'Error attempting to unlock wallet', error, true)
    Logger.showError(i18n.t(ErrorMessages.ACCOUNT_UNLOCK_FAILED))
    return false
  }
}

export async function removeAccountLocally(account: string) {
  clearPasswordCaches()
  return Promise.all([
    removeStoredItem(STORAGE_KEYS.PEPPER),
    removeStoredItem(passwordHashStorageKey(account)),
  ])
}
