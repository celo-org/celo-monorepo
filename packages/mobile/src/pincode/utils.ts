import { getContractKitOutsideGenerator } from 'src/web3/contracts'
import { readPrivateKeyFromLocalDisk } from 'src/web3/privateKey'

export const PIN_LENGTH = 6

export function isPinValid(pin: string) {
  return pin.length === PIN_LENGTH
}

export function isPinCorrect(
  pin: string,
  fornoMode: boolean,
  currentAccount: string
): Promise<typeof pin> {
  return new Promise((resolve, reject) => {
    if (fornoMode) {
      readPrivateKeyFromLocalDisk(currentAccount, pin)
        .then(() => resolve(pin))
        .catch(reject)
    } else {
      getContractKitOutsideGenerator()
        .then((contractKit: any) =>
          contractKit.web3.eth.personal
            .unlockAccount(currentAccount, pin, 1)
            .then((result: boolean) => (result ? resolve(pin) : reject()))
        )
        .catch(reject)
    }
  })
}
