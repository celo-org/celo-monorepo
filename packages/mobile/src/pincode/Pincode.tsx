/**
 * This is a VIEW. We use it everwhere we need to show PIN pad
 * with an input, e.g. get/ensure/set pincode.
 */
import NumberKeypad from '@celo/react-components/components/NumberKeypad'
import colors from '@celo/react-components/styles/colors.v2'
import fontStyles from '@celo/react-components/styles/fonts.v2'
import React from 'react'
import { ScrollView, StyleSheet, Text, View } from 'react-native'
import PincodeDisplay from 'src/pincode/PincodeDisplay'
import { PIN_LENGTH } from 'src/pincode/utils'

interface Props {
  title?: string
  errorText?: string
  maxLength?: number
  pin: string
  onChangePin: (pin: string) => void
  onCompletePin: (pin: string) => void
}

function Pincode({
  title,
  errorText,
  maxLength = PIN_LENGTH,
  pin,
  onChangePin,
  onCompletePin,
}: Props) {
  function onDigitPress(digit: number) {
    if (pin.length >= maxLength) {
      return
    }

    const newPin = pin + digit
    onChangePin(newPin)
    if (newPin.length === maxLength) {
      // Wait for next frame so we the user can see the last digit
      // displayed before acting on it
      requestAnimationFrame(() => onCompletePin(newPin))
    }
  }

  function onBackspacePress() {
    onChangePin(pin.substr(0, pin.length - 1))
  }

  return (
    <ScrollView style={styles.container} contentContainerStyle={styles.scrollContainer}>
      <View style={styles.spacer} />
      {!errorText && <Text style={styles.title}>{title || ' '}</Text>}
      {!!errorText && <Text style={styles.error}>{errorText}</Text>}
      <View style={styles.pincodeContainer}>
        <PincodeDisplay pin={pin} maxLength={maxLength} />
      </View>
      <View style={styles.spacer} />
      <NumberKeypad onDigitPress={onDigitPress} onBackspacePress={onBackspacePress} />
    </ScrollView>
  )
}

const styles = StyleSheet.create({
  container: {
    marginTop: 40,
  },
  scrollContainer: {
    flexGrow: 1,
  },
  spacer: {
    flex: 1,
  },
  title: {
    ...fontStyles.regular,
    textAlign: 'center',
    marginBottom: 24,
  },
  error: {
    ...fontStyles.regular500,
    color: colors.warning,
    textAlign: 'center',
    marginBottom: 24,
  },
  pincodeContainer: {
    marginBottom: 24,
    paddingHorizontal: '15%',
    alignItems: 'center',
  },
})

export default Pincode
