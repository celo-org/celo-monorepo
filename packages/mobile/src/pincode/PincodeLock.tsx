/**
 * This is a VIEW, which we use as an overlay, when we need
 * to lock the app with a PIN code.
 */
import colors from '@celo/react-components/styles/colors'
import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { BackHandler, StyleSheet } from 'react-native'
import RNExitApp from 'react-native-exit-app'
import { SafeAreaView } from 'react-native-safe-area-context'
import { useDispatch, useSelector } from 'react-redux'
import { appUnlock } from 'src/app/actions'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { Namespaces } from 'src/i18n'
import Pincode from 'src/pincode/Pincode'
import { ensureCorrectPin } from 'src/pincode/utils'
import { currentAccountSelector } from 'src/web3/selectors'

function PincodeLock() {
  const [pin, setPin] = useState('')
  const [errorText, setErrorText] = useState<string | undefined>(undefined)
  const dispatch = useDispatch()
  const { t } = useTranslation(Namespaces.nuxNamePin1)
  const currentAccount = useSelector(currentAccountSelector)

  function onWrongPin() {
    setPin('')
    setErrorText(t(`${Namespaces.global}:${ErrorMessages.INCORRECT_PIN}`))
  }

  function onCorrectPin() {
    dispatch(appUnlock())
  }

  function onCompletePin(enteredPin: string) {
    if (currentAccount) {
      return ensureCorrectPin(pin, currentAccount)
        .then(onCorrectPin)
        .catch(onWrongPin)
    } else {
      onCorrectPin()
    }
  }

  useEffect(() => {
    function hardwareBackPress() {
      RNExitApp.exitApp()
      return true
    }
    const backHandler = BackHandler.addEventListener('hardwareBackPress', hardwareBackPress)
    return function cleanup() {
      backHandler.remove()
    }
  }, [])

  return (
    <SafeAreaView style={style.container}>
      <Pincode
        title={t('confirmPin.title')}
        errorText={errorText}
        pin={pin}
        onChangePin={setPin}
        onCompletePin={onCompletePin}
      />
    </SafeAreaView>
  )
}

const style = StyleSheet.create({
  container: {
    paddingTop: 20,
    flex: 1,
    backgroundColor: colors.background,
  },
})

export default PincodeLock
