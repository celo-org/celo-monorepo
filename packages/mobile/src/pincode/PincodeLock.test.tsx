import * as React from 'react'
import { fireEvent, flushMicrotasksQueue, render } from 'react-native-testing-library'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import { appUnlock } from 'src/app/actions'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { Namespaces } from 'src/i18n'
import PincodeLock from 'src/pincode/PincodeLock'
import { ensureCorrectPin } from 'src/pincode/utils'
import { createMockStore } from 'test/utils'

const pin = '123456'

describe('PincodeLock', () => {
  it('renders correctly', () => {
    const store = createMockStore()
    const tree = renderer.create(
      <Provider store={store}>
        <PincodeLock />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('unlocks if PIN is correct', async () => {
    ;(ensureCorrectPin as jest.Mock).mockResolvedValueOnce(pin)
    const store = createMockStore()

    const { getByTestId } = render(
      <Provider store={store}>
        <PincodeLock />
      </Provider>
    )
    pin.split('').forEach((number) => fireEvent.press(getByTestId(`digit${number}`)))
    jest.runAllTimers()
    await flushMicrotasksQueue()
    expect(store.getActions()).toEqual([appUnlock()])
  })

  it('shows wrong PIN notification', async () => {
    ;(ensureCorrectPin as jest.Mock).mockRejectedValue('')
    const store = createMockStore()

    const { getByTestId, getByText } = render(
      <Provider store={store}>
        <PincodeLock />
      </Provider>
    )
    pin.split('').forEach((number) => fireEvent.press(getByTestId(`digit${number}`)))
    jest.runAllTimers()
    await flushMicrotasksQueue()
    expect(getByText(`${Namespaces.global}:${ErrorMessages.INCORRECT_PIN}`)).toBeDefined()
    expect(store.getActions()).toEqual([])
  })
})
