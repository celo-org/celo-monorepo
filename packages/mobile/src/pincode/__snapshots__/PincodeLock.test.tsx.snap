// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PincodeLock renders correctly 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "paddingTop": 20,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
        "padding": 20,
        "paddingTop": 0,
      }
    }
  >
    <View>
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Light",
                "fontSize": 22,
                "paddingBottom": 20,
                "textAlign": "center",
              },
              Object {
                "marginTop": 15,
              },
            ]
          }
        >
          confirmPin.title
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "marginBottom": 20,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderColor": "#EEEEEE",
                  "borderRadius": 3,
                  "borderWidth": 1,
                },
                Object {
                  "alignItems": "center",
                  "height": 55,
                  "justifyContent": "center",
                  "paddingHorizontal": 20,
                  "paddingVertical": 12,
                  "width": 180,
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "#D1D5D8",
                  "fontFamily": "Hind-Regular",
                  "fontSize": 16,
                  "lineHeight": 24,
                }
              }
            >
              createPin.yourPin
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "paddingHorizontal": 12,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit1"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              1
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit2"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              2
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit3"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              3
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit4"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              4
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit5"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              5
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit6"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              6
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit7"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              7
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit8"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              8
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit9"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              9
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "color": "#2E3338",
                "fontFamily": "Inter-Medium",
                "fontSize": 22,
                "justifyContent": "center",
                "lineHeight": 28,
                "padding": 24,
                "textAlign": "center",
                "width": 64,
              }
            }
          />
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            testID="digit0"
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              0
            </Text>
          </View>
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackgroundBorderless",
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "color": "#2E3338",
                  "fontFamily": "Inter-Medium",
                  "fontSize": 22,
                  "justifyContent": "center",
                  "lineHeight": 28,
                  "padding": 24,
                  "textAlign": "center",
                  "width": 64,
                }
              }
            >
              <svg
                fill="none"
                height={30}
                style={Object {}}
                viewBox="0 0 35 18"
                width={30}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1.82198 9.7422L9.57436 16.7422C9.75813 16.9081 9.99693 17 10.2445 17H33C33.5523 17 34 16.5523 34 16V2C34 1.44772 33.5523 1 33 1H10.2445C9.99693 1 9.75813 1.09186 9.57436 1.25779L1.82198 8.25779C1.38221 8.65488 1.38221 9.34512 1.82198 9.7422Z"
                  stroke="#2E3338"
                  strokeWidth="3"
                  style={Object {}}
                />
                <path
                  d="M17 5L25 13M17 13L25 5"
                  stroke="#2E3338"
                  strokeWidth="3"
                  style={Object {}}
                />
              </svg>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 3,
          "flexDirection": "row",
        },
        null,
        null,
        undefined,
        Object {
          "backgroundColor": "#A3EBC6",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={false}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          },
          Object {
            "backgroundColor": "#A3EBC6",
          },
          Object {
            "height": 50,
          },
          Object {
            "borderWidth": 0,
          },
        ]
      }
      testID="Pincode-Submit"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Hind-SemiBold",
                "fontSize": 16,
              },
              Object {
                "color": "#FFFFFF",
              },
              Object {
                "paddingLeft": 5,
                "paddingRight": 5,
              },
            ]
          }
        >
          global:submit
        </Text>
      </View>
    </View>
  </View>
</SafeAreaView>
`;
